Jaeger Receiver Config Fixes

joe-elliott opened the pull request:
Fixes #445, #158 

This PR addresses some Jaeger receiver config cleanup as well as makes some breaking changes to the way the config is handled.   See below for details.

**Fixes/Updates**
- Disabled flag is respected per protocol
- Unspecified protocols will no longer be started
- Empty protocol configs can now be specified to start the protocol with defaults. e.g.
  ```
  jaeger:
    protocols:
      grpc:
  ```
- Updated readmes
- Naming and behavior of per protocol Addr/Enabled functions in `trace_reciever.go` has been standardized.
- Added thrift tchannel test to meet code coverage

**Breaking Change**
Changed the way an empty `jaeger:` config is handled.  An empty/default config does not start any jaeger protocols.  Previously it started all three collector protocols.  This is a consequence of not starting unspecified protocols.

pjanotti commented:
Hi @joe-elliott, what is the behavior when empty `jaeger:` is specified? The breaking change is fine as long as that is reported as an error (so it will be clear that the at least one protocol needs to be specified).

pjanotti commented:
LGTM @joe-elliott just need to fix the test 
```terminal
--- FAIL: TestLoadConfig (0.01s)
    config_test.go:37: 
        	Error Trace:	config_test.go:37
        	Error:      	Received unexpected error:
        	            	error reading settings for receiver type "jaeger": must specify at least one protocol when using the Jaeger receiver
        	Test:       	TestLoadConfig
```

joe-elliott commented:
@pjanotti Thanks for the comments.

When starting the Jaeger receiver with an unknown protocol the collector will log:
```
Cannot load configuration: error reading settings for receiver type "jaeger": unknown jaeger protocol <bad proto>
```

When starting the Jaeger receiver with no protocols the collector will log:
```
Cannot load configuration: error reading settings for receiver type "jaeger": must specify at least one protocol when using the Jaeger receiver
```

tigrannajaryan commented on receiver/jaegerreceiver/testdata/config.yaml:
```diff
@@ -1,7 +1,6 @@
   # The following demonstrates initializing the default jaeger receiver.
-  # By default all three protocols (grpc, thrift-http, thrift-tchannel)
-  # are enabled and the default endpoints are specified in factory.go
+  # By default this configuration enables nothing and will serve no jaeger protocols
```
```suggestion
  # By default this configuration enables nothing and will serve no jaeger protocols.
```

tigrannajaryan replied:
Please follow the convention in this file and end comment sentences with periods.

joe-elliott replied:
Done.

tigrannajaryan commented on receiver/jaegerreceiver/config_test.go:
```diff
@@ -110,3 +160,16 @@ func TestLoadConfig(t *testing.T) {
+	factory := &Factory{}
+	factories.Receivers[typeStr] = factory
+	_, err = config.LoadConfigFile(t, path.Join(".", "testdata", "bad_proto_config.yaml"), factories)
+	assert.NotNil(t, err)
```
Would it be better to assert that `err` is a specific error from this list? https://github.com/open-telemetry/opentelemetry-collector/blob/60b03d0d2d503351501291b30836d2126487a741/config/config.go#L43

Should be likely `errUnmarshalErrorOnReceiver`.

joe-elliott replied:
I was unable to assert the type b/c `errUnmarshalErrorOnReceiver` is private to that package.  I went with this.  If there's a better way, please let me know.
```
assert.EqualError(t, err, `error reading settings for receiver type "jaeger": must specify at least one protocol when using the Jaeger receiver`)
```
Should those err types be exported so that consuming packages can better check config failures?

tigrannajaryan replied:
No, I think what you have is fine. No need to export err types.

tigrannajaryan commented on receiver/jaegerreceiver/factory.go:
```diff
@@ -59,33 +64,42 @@ func (f *Factory) Type() string {
 func (f *Factory) CustomUnmarshaler() receiver.CustomUnmarshaler {
-	return nil
+	return func(v *viper.Viper, viperKey string, intoCfg interface{}) error {
+		err := v.UnmarshalKey(viperKey, intoCfg, func(decoderCfg *mapstructure.DecoderConfig) {})
```
This doesn't seem to be an equivalent functionality to what it was before. When there is no customer unmarhaler defined Config uses `UnmarshalExact` which errors if there is an option that does not exist in the struct. `UnmarshalKey` doesn't behave that way.

joe-elliott replied:
This is true.  I chose this b/c `UnmarshalExact` would not add an entry to the map if there were no subkeys. i.e.
```
jaeger:
  protocols:
    grpc:  
```
would not unmarshal to a "grpc" entry in the `receiverCfg.protocols` map where as `UnmarshalKey` did.  I agree this functionality is different.  I will submit a better implementation.

joe-elliott replied:
I'm reviewing options to better reproduce the original functionality along with the custom functionality we need.  @tigrannajaryan are you ok with exporting this method:

https://github.com/open-telemetry/opentelemetry-collector/blob/60b03d0d2d503351501291b30836d2126487a741/config/config.go#L857-L859

from the config package.  There is a lot of custom functionality in here related to expanding env vars I'd have to reproduce in my custom unmarshaller to get the same behavior.  I think this functionality could be useful for future custom unmarshallers as well.

tigrannajaryan replied:
Yes, I think it is OK to export getConfigSection. It implements a generic functionality that we should be able to provide as a public API for components that need custom unmarshaling. Let's see the implementation you have for customer unmarshaler.

tigrannajaryan commented on receiver/jaegerreceiver/trace_receiver.go:
```diff
@@ -442,56 +437,63 @@ func (jr *jReceiver) buildProcessor(address string, factory apacheThrift.TProtoc
-	if terr != nil {
-		return fmt.Errorf("failed to create NewTChannel: %v", terr)
+	if !jr.collectorGRPCEnabled() && !jr.collectorHTTPEnabled() && !jr.collectorThriftEnabled() {
+		return nil
```
Shouldn't this return error since nothing is enabled?

joe-elliott replied:
In `StartTraceReception()` both the collector and agent protocols are started.  A receiver could be configured for only agent protocols which would be a valid configuration but still return nil here.

https://github.com/open-telemetry/opentelemetry-collector/blob/39b9ea98ccc9b2e8cc547fa434ece6793233edf4/receiver/jaegerreceiver/trace_receiver.go#L212-L220

I had considered breaking each protocol into it's own "start" function instead of `startAgent()` and `startCollector()`, but I thought that might be too much churn for this PR.

tigrannajaryan commented on receiver/jaegerreceiver/trace_receiver.go:
```diff
@@ -442,56 +437,63 @@ func (jr *jReceiver) buildProcessor(address string, factory apacheThrift.TProtoc
+		caddr := jr.collectorHTTPAddr()
+		cln, cerr := net.Listen("tcp", caddr)
+		if cerr != nil {
+			// Abort and close tch
```
How is `tch` closed?

joe-elliott replied:
If `startCollector()` fails then `StartTraceReception()` calls `stopTraceReceptionLocked()`.  This will clean up tchannel or any other open services.

Removing the comment b/c it's misleading.

tigrannajaryan commented on receiver/jaegerreceiver/trace_receiver_test.go:
```diff
@@ -101,6 +106,41 @@ func TestReception(t *testing.T) {
+	assert.NoError(t, err, "should not have failed to start trace reception")
+
+	// there is a race condition here that we're ignoring.
+	//   this test may occasionally pass incorrectly, but it will not fail incorrectly
```
Does this realistically ever lose the race and verify anything? Unless you introduce a deliberate pause my guess would be this almost always executed before StartTraceReception has a chance to listen. I could be wrong, let me know if you tested for this.

Side note: we should probably allow components to asynchronously signal end of Start so that tests can wait for that.

joe-elliott replied:
I added this test and verified it failed consistently before adding the fix.  Given that it was consistently failing I chose not to add a pause to not add anymore time to CI.  I would be glad to add a pause (or even remove the test) if you'd like.

I agree on your side note.  There's a number of tests where we have a "WaitForPort()" call.  A direct signal would be much cleaner.

tigrannajaryan replied:
If it fails consistently or even most of the time then it is good enough for now. We can revisit the test when we have better signalling of completion from components to host. Please add a TODO to the comment to reflect this discussion.

joe-elliott commented:
@tigrannajaryan  I believe I have addressed your concerns. 

Note that I went ahead and exported `GetConfigSection` from the config package.  It seems like this function will be useful for any future custom unmarshallers as it will allow them to easily reuse all of the environment substitution logic in the config package.  If this is not desirable, I will revert and try a different path.

tigrannajaryan commented on receiver/jaegerreceiver/factory.go:
```diff
@@ -59,33 +64,49 @@ func (f *Factory) Type() string {
-	return nil
+	return func(v *viper.Viper, viperKey string, intoCfg interface{}) error {
+		// first load the config normally
+		vReceiverCfg := config.GetConfigSection(v, viperKey)
```
I thought a bit more about this and there is an alternate solution.

Instead of making GetConfigSection exported and calling it here for the second time, what if we simply pass to CustomUnmarshaler one more parameter, the viper section that was already fetched by the earlier call to GetConfigSection in loadReceivers, and name the parameter e.g. `sourceViperSection *viper.Viper`?

This way there is no longer need to export GetConfigSection, don't need to call it the second time and do it in a way that matches the behavior of loadReceivers, which is and internal and may be changed in the future in a way that we cannot predict now.

What do you think?

joe-elliott replied:
Yup, works for me.

I had considered something like this since `sv` is sitting right there in `loadReceivers()` and is exactly what I needed.  Will push the fix shortly.

tigrannajaryan commented on receiver/jaegerreceiver/factory.go:
```diff
@@ -59,33 +63,48 @@ func (f *Factory) Type() string {
+		if vSub == nil {
+			return fmt.Errorf("Jaeger receiver config is empty")
+		}
+		protocols := vSub.GetStringMap("protocols")
```
I think it will be nice to define "protocols" as a const somewhere in `config.go` near `Config struct` to make it clear that this is the key of `Protocols` field and corresponds to the mapstructure tag.

tigrannajaryan commented:
Mostly LGTM, only a couple small nits.

joe-elliott commented:
@tigrannajaryan I believe that covers it.  Let me know if I can make any other changes.
